sc-kpm/sc-search/agents/GetDecompositionAgent.cpp

```diff
@@ -0,0 +1,139 @@
+ * Distributed under the MIT License
+ * (See accompanying file COPYING.MIT or copy at http://opensource.org/licenses/MIT)
+ */
+
+#include "sc-agents-common/keynodes/coreKeynodes.hpp"
+#include "sc-agents-common/utils/AgentUtils.hpp"
+#include "sc-agents-common/utils/CommonUtils.hpp"
+#include "sc-agents-common/utils/IteratorUtils.hpp"
+
+#include "constants/ScSearchAliases.hpp"
+#include "keynodes/Keynodes.hpp"
+
+#include "GetDecompositionAgent.hpp"
+
+using namespace utils;
+
+namespace scSearch
+{
+SC_AGENT_IMPLEMENTATION(GetDecompositionAgent)
+{
+  ScAddr questionNode = m_memoryCtx.GetEdgeTarget(edgeAddr);
+  if (!checkActionClass(questionNode))
+    return SC_RESULT_OK;
+
+  SC_LOG_DEBUG("GetDecompositionAgent started");
+
+  int level = 1;
+  ScAddr subjDomainAddr = IteratorUtils::getAnyByOutRelation(
+        &m_memoryCtx, questionNode, scAgentsCommon::CoreKeynodes::rrel_1);
+  ScAddr levelAddr = IteratorUtils::getAnyByOutRelation(
+        &m_memoryCtx, questionNode, scAgentsCommon::CoreKeynodes::rrel_2);
+  ScAddr langAddr = IteratorUtils::getAnyByOutRelation(
+        &m_memoryCtx, questionNode, scAgentsCommon::CoreKeynodes::rrel_3);
+
+  if (!m_memoryCtx.IsElement(subjDomainAddr))
+  {
+    SC_LOG_ERROR("GetDecompositionAgent: subject domain node not found.");
+    utils::AgentUtils::finishAgentWork(&m_memoryCtx, questionNode, false);
+    return SC_RESULT_ERROR_INVALID_PARAMS;
+  }
+  if (m_memoryCtx.IsElement(levelAddr))
+    level = stoi(CommonUtils::getLinkContent(&m_memoryCtx, levelAddr));
+  if (!m_memoryCtx.IsElement(langAddr))
+  {
+    SC_LOG_ERROR("GetDecompositionAgent: language node not found.");
+    utils::AgentUtils::finishAgentWork(&m_memoryCtx, questionNode, false);
+    return SC_RESULT_ERROR_INVALID_PARAMS;
+  }
+
+  ScAddr answerLink = m_memoryCtx.CreateLink();
+
+  ScAddrVector decomposition = getDecomposition(subjDomainAddr);
+  json answerJSON {
+        {CommonUtils::getScWebAddr(subjDomainAddr), {
+              {"idtf", CommonUtils::getMainIdtf(&m_memoryCtx, subjDomainAddr, {langAddr})},
+              {"decomposition", getJSONDecomposition(decomposition, level, langAddr)},
+              {"position", 0}
+        }}
+  };
+
+  std::string answerJSONContent = answerJSON.dump();
+  SC_LOG_DEBUG(answerJSONContent);
+  m_memoryCtx.SetLinkContent(answerLink, answerJSONContent);
+
+  ScAddrVector answerElements = {answerLink};
+
+  utils::AgentUtils::finishAgentWork(&m_memoryCtx, questionNode, answerElements, true);
+  SC_LOG_DEBUG("GetDecompositionAgent finished");
+  return SC_RESULT_OK;
+}
+
+bool GetDecompositionAgent::checkActionClass(ScAddr const & actionNode)
+{
+  return m_memoryCtx.HelperCheckEdge(Keynodes::action_get_decomposition, actionNode, ScType::EdgeAccessConstPosPerm);
+}
+
+ScAddrVector GetDecompositionAgent::getDecomposition(ScAddr const & subjDomainAddr)
+{
+  SC_LOG_DEBUG("Main section: " + ms_context->HelperGetSystemIdtf(subjDomainAddr));
+  ScAddrVector decomposition;
```

Comment:
Can we use ScAddrList here?

Reply:
This function returns vector which is goint to be used to generate json.